### PR TITLE
Core Mapping - Measuring Tool

### LINKED ISSUE #2868: Core Mapping - Measuring Tool (basic)
The core mapping could do with these two features:
- ScaleBar - to show the current scale (in km).
- Measuring Tool - switch on to allow the user to measure a single or poly line. We should add this as a control to the geoman toolbar

We've used both in the past, we can probably use similar solutions this time.

### PR DESCRIPTION
**Pull request by lebaphi**
Fixes #2868

### CODE REVIEW

**Comment by lebaphi** [comment]
image

**Comment by lebaphi** [comment]
it does not work in storybok, I'm working on it

**Comment by IanMayo** [comment]
Trouble in deployment for game @lebaphi :
image

Storybook broken too:
image

I guess you haven't added the ruler control to the repo.

**Comment by Tristina1788** [comment]
@IanMayo , @lebaphi  Testing from '2868-measuring-tool', I get the white screen when access from url localhost:8080, but it works fine on localhost:3000


https://github.com/serge-web/serge/assets/107697044/ac797fec-92f6-4def-84a9-084156a65e14

- Measuring tool is still enable when I select another function (such as : create new shape item).

https://github.com/serge-web/serge/assets/107697044/d668c5ec-2add-48e0-90fb-635b49ec584e

- Measuring icon should change to disable icon when I click on icon ( for disable) instead of when I move the pointer out of icon.


https://github.com/serge-web/serge/assets/107697044/a503c731-2f03-4cac-8cde-9afb2821f20c

**Comment by lebaphi** [comment]
> Trouble in deployment for game @lebaphi : image
> 
> Storybook broken too: image
> 
> I guess you haven't added the ruler control to the repo.

Hi @IanMayo , I think the ruler is already in the repo 
image

**Comment by IanMayo** [comment]
> Hi @IanMayo , I think the ruler is already in the repo

Hmm, something else must be causing the Heroku deployment (and storybook) to fail.

**Comment by IanMayo** [comment]
That's strange @lebaphi    At the Heroku URL for this PR, it is still breaking:
https://serge-2868-measuring-to-a7jsmx.herokuapp.com/?wargame=wargame-l6nngxlk&access=blue-mtg-1

**Comment by lebaphi** [comment]
> > Hi @IanMayo , I think the ruler is already in the repo
> 
> Hmm, something else must be causing the Heroku deployment (and storybook) to fail.

it seems to work locally (develop & storybook), I'm not sure why it fails on heroku 🤔

**Comment by IanMayo** [comment]
Yes, it works locally for storybook and :3000.  We won't be able to review future PRs on Heroku if it's crashing though.

**Comment by Tristina1788** [comment]
@lebaphi , @IanMayo  Tested on latest code '2868-measuring-tool' branch, now measuring tool can works as expected. Only issue on localhost:8080 still happens


https://github.com/serge-web/serge/assets/107697044/fe2e77e7-4b97-41a1-bb40-7e0bbede086e

Core Mapping - Measuring Tool 8080

**Comment by Tristina1788** [comment]
@lebaphi Check on latest code on '2868-measuring-tool' branch, I still get white screen on http://localhost:8080/

image

**Comment by lebaphi** [comment]
Hi @Tristina1788, I'm still working on it, I'll let you know when it's ready for testing

**Comment by lebaphi** [comment]
it works but I think it's not a good solution, it should be fixes in `vite.config.ts` I think, need to investigate more.

**Comment by IanMayo** [comment]
> it works but I think it's not a good solution, it should be fixes in `vite.config.ts` I think, need to investigate more.

Sure, ok.

**Comment by IanMayo** [comment]
> it works but I think it's not a good solution, it should be fixes in `vite.config.ts` I think, need to investigate more.

Sure, it does look scary @lebaphi 

image

**Comment by IanMayo** [comment]
Could you retest please @Tristina1788

**Comment by Tristina1788** [comment]
@IanMayo , @lebaphi  Now it can works on http://localhost:8080/

image

- But It found an issue. Measuring line is not clean when I click another function on map. After that I can't enable measuring tool to measure line. (Need to refresh page to use measuring tool)


https://github.com/serge-web/serge/assets/107697044/669b8b0b-9e3a-4767-8e35-3bb5e1b5c0a2

- And we can't use measuring tool with start point is a point in shape. Do you think we have that case or not?

https://github.com/serge-web/serge/assets/107697044/cb4a248b-9d67-4a38-91d0-ca4dc880782f

**Comment by IanMayo** [comment]
Yes, @tristina - I guess we need to prevent map-select events when the measuring tool is active.  Can you investigate that please @lebaphi

**Comment by Tristina1788** [comment]
@lebaphi , @IanMayo  Now  I can use measuring tool with start point is a point in shape.  And measuring line is not showed when select another function as expected


https://github.com/serge-web/serge/assets/107697044/8dcdf8e5-dd33-4944-88f6-0b6b3dc797b5

But sometimes I have to click 2,3 times to click point on measuring tool. @lebaphi , @IanMayo do you see it when use measuring tool?



https://github.com/serge-web/serge/assets/107697044/6a45861e-7816-4b13-984c-535b126c7db2

File: client/static/leaflet/ruler/leaflet.ruler.js [hunk 1]
```diff
@@ -54,6 +54,7 @@
       this._clickedLatLong = null
       this._clickedPoints = []
       this._totalLength = 0
```
Comment (IanMayo) [inline client/static/leaflet/ruler/leaflet.ruler.js:56]: we can drop this next debug line, right @lebaphi ?

Comment (lebaphi) [reply client/static/leaflet/ruler/leaflet.ruler.js:56]: yep we can debug this class in devtool @IanMayo

Comment (IanMayo) [reply client/static/leaflet/ruler/leaflet.ruler.js:56]: Stupid GH comments 🤣 

I was referring to the console.log on line 57

Comment (lebaphi) [reply client/static/leaflet/ruler/leaflet.ruler.js:56]: ah I see 🤣🤣

**Comment by IanMayo** [comment]
> But sometimes I have to click 2,3 times to click point on measuring tool.  @IanMayo do you see it when use measuring tool?

I wasn't able to reproduce the issue, but I now have. If I repeatedly click really quickly, the points aren't picked up:

screencast

I don't think this means the feature is _broken_, but it could still be _annoying_.  Could you investigate please @lebaphi - I wonder if it's associated with the introduction of the `delay` command?

**Comment by lebaphi** [comment]
Hi @IanMayo, I can't reproduce this issue either, `delay` command use to set measurement status (true or false), if true we will prevent clicking on shape/marker, it unrelated to measurement behavior.
image

**Comment by Tristina1788** [comment]
I checked the latest code and everything is good  except this 'sometimes I have to click 2,3 times to click point on measuring tool'. It's still happened on my side.

**Comment by Tristina1788** [comment]
I try to test on Firefox (instead of Chrome as before). And It's much better. I don't know if the issue related to cache or browser.

**Comment by IanMayo** [comment]
Sure, ok @Tristina1788 

In that case, I suggest we "live with it" for now, and see if a Chrome update fixes it in the near future.

**Review by IanMayo — APPROVED** [review]

**Pull request merged**